lapis/src/main/kotlin/org/genspectrum/lapis/model/AdvancedQueryCustomListener.kt

```diff
@@ -0,0 +1,516 @@
+        throw BadRequestException("Invalid nucleotide symbol: $c")
+    }
+}
+
+fun isValidNucleotideSymbol(c: Char) {
+    if (c.uppercaseChar() !in ambiguousNucSymbols && c.uppercaseChar() !in nucleotideSymbols) {
+        throw BadRequestException("Invalid nucleotide symbol: $c")
+    }
+}
+
+fun isValidAminoAcidSymbol(c: Char) {
+    if (c.uppercaseChar() !in ambiguousAaSymbols && c.uppercaseChar() !in aaSymbols) {
+        throw BadRequestException("Invalid amino acid symbol: $c")
+    }
+}
+
+class AdvancedQueryCustomListener(
+    val referenceGenomeSchema: ReferenceGenomeSchema,
+    databaseConfig: DatabaseConfig,
+) : AdvancedQueryBaseListener(),
+    ParseTreeListener {
+    private val expressionStack = ArrayDeque<SiloFilterExpression>()
+
+    private val metadataFieldsByName = databaseConfig.schema.metadata
+        .associateBy { it.name.lowercase(Locale.US) }
+
+    fun getAdvancedQueryExpression(): SiloFilterExpression {
+        if (expressionStack.size != 1) {
+            log.error { "Expected exactly one expression on the stack, but got this stack $expressionStack." }
+            throw RuntimeException("Failed to parse advanced query.")
+        }
+
+        return expressionStack.first()
+    }
+
+    override fun enterMetadataLessThanEqualQuery(ctx: AdvancedQueryParser.MetadataLessThanEqualQueryContext) {
+        val metadataName = ctx.geneOrName()[0].text
+        val metadataValue = ctx.geneOrName()[1].text
+
+        val field: DatabaseMetadata =
+            metadataFieldsByName[metadataName.lowercase(Locale.US)]
+                ?: throw BadRequestException("Metadata field $metadataName does not exist", null)
+        when (field.type) {
+            MetadataType.DATE -> {
+                try {
+                    val date = LocalDate.parse(metadataValue)
+                    expressionStack.addLast(DateBetween(field.name, from = null, to = date))
+                } catch (exception: DateTimeParseException) {
+                    throw BadRequestException("'$metadataValue' is not a valid date: ${exception.message}", exception)
+                }
+            }
+
+            MetadataType.FLOAT -> {
+                try {
+                    expressionStack.addLast(FloatBetween(field.name, from = null, to = metadataValue.toDouble()))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid float", e)
+                }
+            }
+
+            MetadataType.INT -> {
+                try {
+                    expressionStack.addLast(IntBetween(field.name, to = metadataValue.toInt(), from = null))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid integer", e)
+                }
+            }
+
+            else -> {
+                throw BadRequestException("expression <= cannot be used for ${field.type}", null)
```

Comment:
I think it would be good to tell the user the name of the field to help in long queries
```suggestion
                throw BadRequestException("expression <= cannot be used for field '${field.name}' of type ${field.type}")
```

(also for the other functions where we have this kind of error)